core/intrinsics/intrinsics.odin

```diff
@@ -162,7 +162,12 @@ type_is_matrix           :: proc($T: typeid) -> bool ---
 type_has_nil :: proc($T: typeid) -> bool ---
 
 type_is_specialization_of :: proc($T, $S: typeid) -> bool ---
+
 type_is_variant_of :: proc($U, $V: typeid) -> bool where type_is_union(U) ---
+type_union_tag :: proc($U: typeid) -> typeid where type_is_union(U) ---
```

Comment:
Is perhaps the name `type_union_tag` weird? Sounds like it would return a tag (which I think of as an int), but it returns a typeid.

Reply:
(I talked to jakub separately about this and I got my confusion fixed)

Reply:
Yep, but maybe a better name is something like `type_union_tag_type`, similar to `type_field_type` intrinsic for getting struct field type by name.

Reply:
I was just about to comment that `tag_type` souns even weirder: This name suggests that the _type_ of the tag is returned, rather than the value of the tag. Maybe `type_union_tag_value` returning typeid is a better option?

Reply:
But the type of the tag _is_ returned :). You can get the variant index with `type_variant_index_of`, and the tag with `min(U) + type_variant_index_of(U, V)`

Reply:
The variant index/tag is an integer, however `type_union_tag_type` returns compile-time typeid constant.

Reply:
Can you help me understand it, let's say in a union you brought up as an example:

```
Player_State :: union { Walking, Flying }
```

What are the tag types and values in this union?

Reply:
The tag type could be u8, u16, u32, or u64, depending on the number of variants within the union, and the required alignment.

The variant values are like values in enums, so Walking would have index 0 (but tag 1), and Flying index 1 (but tag 2).
The tag starts at 1, since 0 is reserved for nil, unless you use `#no_nil`. The reason why the intrinsics return variant index and not the tag is to make both nil/no_nil consistent.

A better example would be something like the following:
```odin
// size = 8 bytes (4 bytes data, 4 aligned variant index)
Foo :: union {
	u8, // index 0, tag 1
	f32, // index 1, tag 2
	i32, // index 2, tag 3
}

Raw_Foo :: struct {
	using _:       struct #raw_union {
		variant_0: u8,
		variant_1: f32,
		variant_2: i32,
	},
	variant_index: u32,
}

// The following is true:

type_union_tag_type(Foo) == u32
type_union_tag_offset(Foo) == 4

type_variant_index_of(Foo, u8) == 0
type_variant_index_of(Foo, i32) == 2
type_variant_type_of(Foo, 0) == u8
...
```

Reply:
Thanks for clarification, I think I misunderstood how unions worked